File: lang/csharp/cxlib/BinaryDecoder.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Text;

namespace CX;

/// <summary>
/// Decoder for the compact binary wire format produced by cx_to_ast_bin and
/// cx_to_events_bin.  All integers are little-endian.
/// </summary>
public static class BinaryDecoder
{
    // ── public entry points ───────────────────────────────────────────────────

    /// <summary>Decode a binary AST payload (the bytes AFTER the 4-byte length prefix)
    /// into a <see cref="CXDocument"/>.</summary>
    public static CXDocument DecodeAST(byte[] data)
    {
        var r = new BufReader(data);
        r.U8(); // version byte — currently 1, reserved for future use

        int prologCount = r.U16();
        var prolog = new List<Node>(prologCount);
        for (int i = 0; i < prologCount; i++)
            prolog.Add(ReadNode(r));

        int elemCount = r.U16();
        var elements = new List<Node>(elemCount);
        for (int i = 0; i < elemCount; i++)
            elements.Add(ReadNode(r));

        return new CXDocument { Prolog = prolog, Elements = elements };
    }

    /// <summary>Decode a binary events payload (the bytes AFTER the 4-byte length prefix)
    /// into a list of <see cref="StreamEvent"/>.</summary>
    public static List<StreamEvent> DecodeEvents(byte[] data)
    {
        var r = new BufReader(data);
        uint count = r.U32();
        var events = new List<StreamEvent>((int)count);

        for (uint i = 0; i < count; i++)
        {
            byte tid = r.U8();
            var ev = new StreamEvent { Type = EventTypeName(tid) };

            switch (tid)
            {
                case 0x01: // StartDoc — no payload
                case 0x02: // EndDoc  — no payload
                    break;

                case 0x03: // StartElement
                    ev.Name     = r.Str();
                    ev.Anchor   = r.OptStr();
                    ev.DataType = r.OptStr();
                    ev.Merge    = r.OptStr();
                    ev.Attrs    = ReadAttrs(r, r.U16());
                    break;

                case 0x04: // EndElement
                    ev.Name = r.Str();
                    break;

                case 0x05: // Text
                case 0x07: // Comment
                case 0x0A: // RawText
                    ev.Value = r.Str();
                    break;

                case 0x06: // Scalar
                {
                    string dt = r.Str();
                    ev.DataType = dt;
                    ev.Value = Coerce(dt, r.Str());
                    break;
                }

                case 0x08: // PI
                    ev.Target = r.Str();
                    ev.Data   = r.OptStr();
                    break;

                case 0x09: // EntityRef
                case 0x0B: // Alias
                    ev.Value = r.Str();
                    break;
            }

            events.Add(ev);
        }

        return events;
    }

    // ── AST node reader ───────────────────────────────────────────────────────

    private static Node ReadNode(BufReader r)
    {
        byte tid = r.U8();

        switch (tid)
        {
            case 0x01: // Element
            {
                string name   = r.Str();
                string? anchor = r.OptStr();
                string? dt    = r.OptStr();
                string? merge = r.OptStr();
                var attrs     = ReadAttrs(r, r.U16());
                int childCount = r.U16();
                var items = new List<Node>(childCount);
                for (int i = 0; i < childCount; i++)
                    items.Add(ReadNode(r));
                return new Element(name)
                {
                    Anchor   = anchor,
                    DataType = dt,
                    Merge    = merge,
                    Attrs    = attrs,
                    Items    = items,
                };
            }

            case 0x02: // Text
                return new TextNode(r.Str());

            case 0x03: // Scalar
            {
                string dt = r.Str();
                return new ScalarNode(dt, Coerce(dt, r.Str()));
            }

            case 0x04: // Comment
                return new CommentNode(r.Str());

            case 0x05: // RawText
                return new RawTextNode(r.Str());

            case 0x06: // EntityRef
                return new EntityRefNode(r.Str());

            case 0x07: // Alias
                return new AliasNode(r.Str());

            case 0x08: // PI
            {
                string target = r.Str();
                string? data  = r.OptStr();
                return new PINode(target, data);
            }

            case 0x09: // XMLDecl
            {
                string ver       = r.Str();
                string? encoding = r.OptStr();
                string? sa       = r.OptStr();
                return new XMLDeclNode(ver, encoding, sa);
            }

            case 0x0A: // CXDirective
                return new CXDirectiveNode(ReadAttrs(r, r.U16()));

            case 0x0C: // BlockContent
            {
                int childCount = r.U16();
                var items = new List<Node>(childCount);
                for (int i = 0; i < childCount; i++)
                    items.Add(ReadNode(r));
                return new BlockContentNode(items);
            }

            case 0xFF: // skip / DTD placeholder — no payload
                return new TextNode("");

            default:
                // Unknown type: we cannot safely skip without knowing payload size.
                // Return an empty text node as a best-effort fallback.
                return new TextNode("");
        }
    }

    // ── helpers ───────────────────────────────────────────────────────────────

    private static List<Attr> ReadAttrs(BufReader r, int count)
    {
        var attrs = new List<Attr>(count);
        for (int i = 0; i < count; i++)
        {
            string name     = r.Str();
            string valueStr = r.Str();
            string typeStr  = r.Str();
            string? dt = typeStr == "string" ? null : typeStr;
            attrs.Add(new Attr(name, Coerce(typeStr, valueStr), dt));
        }
        return attrs;
    }

    private static object? Coerce(string typeStr, string valueStr) => typeStr switch
    {
        "int"   => (object?)long.Parse(valueStr,
                        System.Globalization.NumberStyles.Integer,
                        System.Globalization.CultureInfo.InvariantCulture),
        "float" => double.Parse(valueStr,
                        System.Globalization.NumberStyles.Float,
                        System.Globalization.CultureInfo.InvariantCulture),
        "bool"  => valueStr == "true",
        "null"  => null,
        _       => valueStr,
    };

    private static string EventTypeName(byte tid) => tid switch
    {
        0x01 => "StartDoc",
        0x02 => "EndDoc",
        0x03 => "StartElement",
        0x04 => "EndElement",
        0x05 => "Text",
        0x06 => "Scalar",
        0x07 => "Comment",
        0x08 => "PI",
        0x09 => "EntityRef",
        0x0A => "RawText",
        0x0B => "Alias",
        _    => "Unknown",
    };

    // ── BufReader ─────────────────────────────────────────────────────────────

    /// <summary>Cursor-based little-endian reader over a byte array.</summary>
    private sealed class BufReader
    {
        private readonly byte[] _data;
        private int _pos;

        public BufReader(byte[] data) { _data = data; _pos = 0; }

        public byte U8() => _data[_pos++];

        public ushort U16()
        {
            // Use explicit LE decoding to be safe on all platforms.
            ushort v = (ushort)(_data[_pos] | (_data[_pos + 1] << 8));
            _pos += 2;
            return v;
        }

        public uint U32()
        {
            uint v = (uint)(_data[_pos]
                | (_data[_pos + 1] << 8)
                | (_data[_pos + 2] << 16)
                | (_data[_pos + 3] << 24));
            _pos += 4;
            return v;
        }

        public string Str()
        {
            int len = (int)U32();
            string s = Encoding.UTF8.GetString(_data, _pos, len);
            _pos += len;
            return s;
        }

        public string? OptStr()
        {
            byte flag = U8();
            if (flag == 0) return null;
            return Str();
        }
    }
}

File: lang/swift/cxlib/Sources/CXLib/BinaryDecoder.swift
import Foundation

// ── StreamEvent ───────────────────────────────────────────────────────────────

/// A single event from the CX streaming parser.
public struct StreamEvent {
    public var type: String
    // StartElement / EndElement
    public var name: String?
    public var anchor: String?
    public var dataType: String?
    public var merge: String?
    public var attrs: [Attr] = []
    // Text / Comment / RawText / EntityRef / Alias / Scalar value
    public var value: Any?
    // PI
    public var target: String?
    public var data: String?

    public init(type: String) { self.type = type }

    public func isStartElement(_ named: String? = nil) -> Bool {
        type == "StartElement" && (named == nil || name == named)
    }
    public func isEndElement(_ named: String? = nil) -> Bool {
        type == "EndElement" && (named == nil || name == named)
    }
}

// ── BufReader ─────────────────────────────────────────────────────────────────

private struct BufReader {
    let bytes: [UInt8]
    var pos: Int = 0

    init(_ data: Data) {
        self.bytes = Array(data)
    }

    mutating func u8() throws -> UInt8 {
        guard pos < bytes.count else { throw CXError.parse("binary: unexpected end of buffer (u8)") }
        let v = bytes[pos]; pos += 1; return v
    }

    mutating func u16() throws -> UInt16 {
        guard pos + 1 < bytes.count else { throw CXError.parse("binary: unexpected end of buffer (u16)") }
        let lo = UInt16(bytes[pos])
        let hi = UInt16(bytes[pos + 1])
        pos += 2
        return lo | (hi << 8)
    }

    mutating func u32() throws -> UInt32 {
        guard pos + 3 < bytes.count else { throw CXError.parse("binary: unexpected end of buffer (u32)") }
        let a = UInt32(bytes[pos])
        let b = UInt32(bytes[pos + 1])
        let c = UInt32(bytes[pos + 2])
        let d = UInt32(bytes[pos + 3])
        pos += 4
        return a | (b << 8) | (c << 16) | (d << 24)
    }

    mutating func str() throws -> String {
        let len = Int(try u32())
        guard pos + len <= bytes.count else { throw CXError.parse("binary: string overflows buffer") }
        let slice = bytes[pos ..< pos + len]
        pos += len
        guard let s = String(bytes: slice, encoding: .utf8) else {
            throw CXError.parse("binary: invalid UTF-8 in string")
        }
        return s
    }

    mutating func optStr() throws -> String? {
        let flag = try u8()
        guard flag != 0 else { return nil }
        return try str()
    }
}

// ── coercion ──────────────────────────────────────────────────────────────────

private func _coerce(_ typeStr: String, _ valueStr: String) -> Any? {
    switch typeStr {
    case "int":    return Int(valueStr) as Any?
    case "float":  return Double(valueStr) as Any?
    case "bool":   return (valueStr == "true") as Any?
    case "null":   return nil
    default:       return valueStr as Any?   // string / date / datetime / etc.
    }
}

// ── AST decoder ───────────────────────────────────────────────────────────────

private func _readAttr(_ b: inout BufReader) throws -> Attr {
    let name     = try b.str()
    let valueStr = try b.str()
    let typeStr  = try b.str()
    let dt: String? = typeStr == "string" ? nil : typeStr
    return Attr(name, _coerce(typeStr, valueStr), dataType: dt)
}

private func _readNode(_ b: inout BufReader) throws -> Node {
    let tid = try b.u8()
    switch tid {
    case 0x01:  // Element
        let name   = try b.str()
        let anchor = try b.optStr()
        let dt     = try b.optStr()
        let merge  = try b.optStr()
        let attrCount = Int(try b.u16())
        var attrs: [Attr] = []
        attrs.reserveCapacity(attrCount)
        for _ in 0 ..< attrCount { attrs.append(try _readAttr(&b)) }
        let childCount = Int(try b.u16())
        var items: [Node] = []
        items.reserveCapacity(childCount)
        for _ in 0 ..< childCount { items.append(try _readNode(&b)) }
        let el = Element(name, attrs: attrs, items: items)
        el.anchor   = anchor
        el.dataType = dt
        el.merge    = merge
        return .element(el)

    case 0x02:  // Text
        return .text(try b.str())

    case 0x03:  // Scalar
        let typeStr  = try b.str()
        let valueStr = try b.str()
        return .scalar(dataType: typeStr, value: _coerce(typeStr, valueStr))

    case 0x04:  // Comment
        return .comment(try b.str())

    case 0x05:  // RawText
        return .rawText(try b.str())

    case 0x06:  // EntityRef
        return .entityRef(try b.str())

    case 0x07:  // Alias
        return .alias(try b.str())

    case 0x08:  // PI
        let target = try b.str()
        let data   = try b.optStr()
        return .pi(target: target, data: data)

    case 0x09:  // XMLDecl
        let version    = try b.str()
        let encoding   = try b.optStr()
        let standalone = try b.optStr()
        return .xmlDecl(version: version, encoding: encoding, standalone: standalone)

    case 0x0A:  // CXDirective
        let count = Int(try b.u16())
        var attrs: [Attr] = []
        attrs.reserveCapacity(count)
        for _ in 0 ..< count { attrs.append(try _readAttr(&b)) }
        return .cxDirective(attrs)

    case 0x0C:  // BlockContent
        let count = Int(try b.u16())
        var items: [Node] = []
        items.reserveCapacity(count)
        for _ in 0 ..< count { items.append(try _readNode(&b)) }
        return .blockContent(items)

    case 0xFF:  // skip (DTD etc.) — no payload
        return .text("")

    default:
        throw CXError.parse("binary: unknown node type 0x\(String(tid, radix: 16))")
    }
}

// ── Events decoder ────────────────────────────────────────────────────────────

private func _decodeEvents(_ b: inout BufReader) throws -> [StreamEvent] {
    let count = Int(try b.u32())
    var events: [StreamEvent] = []
    events.reserveCapacity(count)

    for _ in 0 ..< count {
        let tid = try b.u8()
        switch tid {
        case 0x01:  // StartDoc
            events.append(StreamEvent(type: "StartDoc"))

        case 0x02:  // EndDoc
            events.append(StreamEvent(type: "EndDoc"))

        case 0x03:  // StartElement
            var e = StreamEvent(type: "StartElement")
            e.name     = try b.str()
            e.anchor   = try b.optStr()
            e.dataType = try b.optStr()
            e.merge    = try b.optStr()
            let attrCount = Int(try b.u16())
            var attrs: [Attr] = []
            attrs.reserveCapacity(attrCount)
            for _ in 0 ..< attrCount {
                let aName    = try b.str()
                let aValStr  = try b.str()
                let aTypeStr = try b.str()
                let dt: String? = aTypeStr == "string" ? nil : aTypeStr
                attrs.append(Attr(aName, _coerce(aTypeStr, aValStr), dataType: dt))
            }
            e.attrs = attrs
            events.append(e)

        case 0x04:  // EndElement
            var e = StreamEvent(type: "EndElement")
            e.name = try b.str()
            events.append(e)

        case 0x05:  // Text
            var e = StreamEvent(type: "Text")
            e.value = try b.str()
            events.append(e)

        case 0x06:  // Scalar
            var e = StreamEvent(type: "Scalar")
            let typeStr  = try b.str()
            let valueStr = try b.str()
            e.dataType = typeStr
            e.value    = _coerce(typeStr, valueStr)
            events.append(e)

        case 0x07:  // Comment
            var e = StreamEvent(type: "Comment")
            e.value = try b.str()
            events.append(e)

        case 0x08:  // PI
            var e = StreamEvent(type: "PI")
            e.target = try b.str()
            e.data   = try b.optStr()
            events.append(e)

        case 0x09:  // EntityRef
            var e = StreamEvent(type: "EntityRef")
            e.value = try b.str()
            events.append(e)

        case 0x0A:  // RawText
            var e = StreamEvent(type: "RawText")
            e.value = try b.str()
            events.append(e)

        case 0x0B:  // Alias
            var e = StreamEvent(type: "Alias")
            e.value = try b.str()
            events.append(e)

        default:
            throw CXError.parse("binary: unknown event type 0x\(String(tid, radix: 16))")
        }
    }
    return events
}

// ── Public API ────────────────────────────────────────────────────────────────

public enum BinaryDecoder {

    /// Decode a binary AST payload (without the 4-byte size prefix) into a CXDocument.
    public static func decodeAST(_ data: Data) throws -> CXDocument {
        var b = BufReader(data)
        _ = try b.u8()  // version byte (currently 1)
        let prologCount = Int(try b.u16())
        var prolog: [Node] = []
        prolog.reserveCapacity(prologCount)
        for _ in 0 ..< prologCount { prolog.append(try _readNode(&b)) }
        let elemCount = Int(try b.u16())
        var elements: [Node] = []
        elements.reserveCapacity(elemCount)
        for _ in 0 ..< elemCount { elements.append(try _readNode(&b)) }
        return CXDocument(elements: elements, prolog: prolog)
    }

    /// Decode a binary events payload (without the 4-byte size prefix) into [StreamEvent].
    public static func decodeEvents(_ data: Data) throws -> [StreamEvent] {
        var b = BufReader(data)
        return try _decodeEvents(&b)
    }
}
